go-support/cloudstate/cloudstate.go

```diff
@@ -0,0 +1,246 @@
+	log.Printf("Received discovery call from sidecar [%s w%s] supporting CloudState %v.%v\n",
+		pi.ProxyName,
+		pi.ProxyVersion,
+		pi.ProtocolMajorVersion,
+		pi.ProtocolMinorVersion,
+	)
+	for _, filename := range []string{
+		"google/protobuf/empty.proto",
+		"google/protobuf/any.proto",
+		"google/protobuf/descriptor.proto",
+		"google/api/annotations.proto",
+		"google/api/http.proto",
+		"cloudstate/event_sourced.proto",
+		"cloudstate/entity.proto",
+		"cloudstate/entity_key.proto",
+	} {
+		if err := r.registerFileDescriptorProto(filename); err != nil {
+			return nil, err
+		}
+	}
+	log.Printf("Responding with: %v\n", r.entitySpec.GetServiceInfo())
+	return r.entitySpec, nil
+}
+
+// ReportError logs
+func (r *EntityDiscoveryResponder) ReportError(c context.Context, fe *protocol.UserFunctionError) (*empty.Empty, error) {
+	log.Printf("ReportError: %v\n", fe)
+	return &empty.Empty{}, nil
+}
+
+func (r *EntityDiscoveryResponder) updateSpec() (err error) {
+	protoBytes, err := proto.Marshal(r.fileDescriptorSet)
+	if err != nil {
+		return errors.New("unable to Marshal FileDescriptorSet")
+	}
+	r.entitySpec.Proto = protoBytes
+	return nil
+}
+
+func (r *EntityDiscoveryResponder) resolveFileDescriptors(dc DescriptorConfig) error {
+	// service
+	if dc.Service != "" {
+		if err := r.registerFileDescriptorProto(dc.Service); err != nil {
+			return err
+		}
+	} else {
+		if dc.ServiceMsg != nil {
+			if err := r.registerFileDescriptor(dc.ServiceMsg); err != nil {
+				return err
+			}
+		}
+	}
+	// and dependent domain descriptors
+	for _, dp := range dc.Domain {
+		if err := r.registerFileDescriptorProto(dp); err != nil {
+			return err
+		}
+	}
+	for _, dm := range dc.DomainMessages {
+		if err := r.registerFileDescriptor(dm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *EntityDiscoveryResponder) registerEntity(e *EventSourcedEntity, config DescriptorConfig) error {
+	if err := r.resolveFileDescriptors(config); err != nil {
+		return fmt.Errorf("failed to resolveFileDescriptor for DescriptorConfig: %+v: %w", config, err)
+	}
+	persistenceID := e.entityName
+	if e.PersistenceID != "" {
+		persistenceID = e.PersistenceID
+	}
+	r.entitySpec.Entities = append(r.entitySpec.Entities, &protocol.Entity{
+		EntityType:    EventSourced,
+		ServiceName:   e.ServiceName,
+		PersistenceId: persistenceID,
+	})
+	// TODO: e.SnapshotEvery
+	return r.updateSpec()
+}
+
+func (r *EntityDiscoveryResponder) registerFileDescriptorProto(filename string) error {
+	descriptorProto, err := unpackFile(proto.FileDescriptor(filename))
+	if err != nil {
+		return fmt.Errorf("failed to registerFileDescriptorProto for filename: %s: %w", filename, err)
+	}
+	r.fileDescriptorSet.File = append(r.fileDescriptorSet.File, descriptorProto)
+	return r.updateSpec()
+}
+
+func (r *EntityDiscoveryResponder) registerFileDescriptor(msg descriptor.Message) error {
+	fd, _ := descriptor.ForMessage(msg) // this can panic, so we do here
```

Comment:
To my regret, I left quite a few Fatal failures left, just so that I got the full attention during start of development for these situations. They're all moved into proper errors with an update to this PR branch.

I also share your concerns regarding exit via `log.Fatalf` or `panic()` a program except for very valid reasons at its startup where it can't start at all. Also libraries should not panic, except perhaps internally.

In this case `descriptor.ForMessage` does panic by itself and also does not visibly states that by its function name, like using MustXXX. So I decided to let it panic until I had a better way to handle it. I fixed it and recover it know and move it into an error `registerFileDescriptor`will return.